### DISCUSSION TITLE
Refactor ride track paint code to move toward paint objects

### DISCUSSION THREAD

**Discussion by spacek531** [OpenRCT2/OpenRCT2#17895] [Feature requests]
I think we're nearing the point where a (admittedly potentially *considerably* large) config file could be used to draw stuff on the screen - there's enough helper functions I think it wouldn't be completely terrible to implement.

Is this something desired?

**Comment by IntelOrca** [comment]
Can you give an example of such a config file?

**Comment by spacek531** [comment]
I envision nested arrays of these things: https://github.com/Gymnasiast/OpenRCT2/blob/97c700b758160befdc6832f3654cd261dd6c2e58/src/openrct2/ride/coaster/WoodenRollerCoaster.h#L18 with some more flags and fields for supports and paint colors etc.

I think the JSON parkobj format is fairly well suited to this kind of thing.

**Reply by IntelOrca** [reply]
It makes sense to have it all declared in a file, but you would need to analyse how all current paint functions work and come up with a suitable schema that allows as many of the different variations of paint as possible and handling all the edge cases.

**Reply by spacek531** [reply]
A lot of the tile elements are simply one sprite being drawn and one support, so a little can go a long way. Special cases can be added as they are needed.

**Comment by frutiemax** [comment]
I can see this being used for the future ride type objects, especially for new track types, and possibly legacy ride types could reuse that paint object to save code duplication.

**Comment by spacek531** [comment]
So far all objects have been selectable by the user on a per-scenario basis, even objects I think maybe shouldn't be like station types. I think it would be nice if users don't have to micro-manage track types - is it reasonable to believe that track types can be always loaded? I think it would be nice if track types could "fall back" on other track types, like the classic wooden coaster and the lim launched coaster which have a few unique sprites and everything else uses the same paint code. It would save having to duplicate a lot of data.

**Comment by duncanspumpkin** [comment]
Yes we call those persistent objects but the idea is that we don't want to get too involved with community objects which lends us towards making the objects not persistent. I'd like to see a new version of test paint for this before a major refactoring then we can be sure we aren't breaking lots.

**Comment by spacek531** [comment]
I have no idea what is involved in test paint. Will you or another dev help with that?

**Reply by duncanspumpkin** [reply]
> I have no idea what is involved in test paint. Will you or another dev help with that?

What it did was place a track then draw the track record what was drawn. Then change the track for each variation (rotation, track type, brake, upside down, ride type, etc.)

**Comment by spacek531** [comment]
Can someone point me to a persistent object? I looked in the object repository but all the objects there looked selectable.

**Reply by duncanspumpkin** [reply]
Audio is currently the only ones like that atm.

**Comment by spacek531** [comment]
As an interim solution to loading objects, I'll define the entries ~~at compile time~~* in the code files so RTDs can use them. *they have a vector which are allergic to constexpr

**Comment by spacek531** [comment]
That is true. I think it can be done incrementally.
